### DISCUSSION TITLE
当前版本（0.36）如何实现声道的左平衡和右平衡

### DISCUSSION THREAD

**Discussion by YSYHZ** [hooke007/mpv_PlayKit#354] [Q&A 问答] [answered]
在[[restore-old-bindings.conf](https://github.com/mpv-player/mpv/blob/master/etc/restore-old-bindings.conf)](url)中可以找到
```
# changed in mpv 0.5.0
( add balance -0.1
) add balance 0.1
```
如何我想要在当前版本通过快捷键调整声道的左平衡和右平衡，该怎么办？
直接在input.conf添加
```
( add balance -0.1
) add balance 0.1
```
报错
![image](https://github.com/hooke007/MPV_lazy/assets/108722788/b95c8c4e-2873-42f0-8072-01e222c75d5e)
Unknown property: 'balance'
未找到balance属性。
在wink上找到了实现相关功能的脚本[https://github.com/wiiaboo/mpv-scripts/blob/master/audio-balance.lua](url)
但是已经是6年前的脚本了，已经失效无法使用

**Comment by hooke007** [comment, marked as answer]
你看那日期就知道早就不维护了没有参考价值。
https://github.com/mpv-player/mpv/blob/master/DOCS/interface-changes.rst 里面提到早就移除了，无替代品

用af=pan做声道平衡 https://ffmpeg.org/ffmpeg-filters.html#pan-1

附局部修复与修改的 [audio-balance.lua](https://gist.github.com/hooke007/8bfe362a45ba20467de057b71fcc084c)

**Reply by hooke007** [reply]
另外说一下 audio-balance.lua 不是失效，就是单纯算错了。。
_多声道不知道怎么算，演示立体声的例子_
正常是这样 `af set pan="stero|FL=1FL|FR=1FR"`
屏蔽左声道就是 `af set pan="stero|FL=0FL|FR=1FR"`

但是脚本里屏蔽左声道是这样 `af set pan="stero|FL=0FL+1FR|FR=1FR"` （把原右声道叠加了进去，屏蔽了个寂寞）

**Reply by YSYHZ** [reply]
老哥救救，看了半天手册了完全看不懂，能教教怎么改audio-balance.lua这个脚本吗？纯小白

**Reply by hooke007** [reply]
我说了 “多声道不知道怎么算”
直接调滤镜用啊

**Reply by YSYHZ** [reply]
我终于看懂了，我用不到多声道直接把audio-balance.lua里关于多声道的全删了。把add_left_channel和add_right_channel这两个函数改了一下就正常工作了

**Reply by hooke007** [reply]
给我看一下你改动的部分，我看原脚本的设计是有混声思路的

**Reply by YSYHZ** [reply]
```
local function add_left_channel(left_ch_name, right_ch_name)
    return string.format("%s=%.1f*%s",
        left_ch_name,
        math.max(0,math.min(1,left*2)), left_ch_name)
end

local function add_right_channel(left_ch_name, right_ch_name)
    return string.format("%s=%.1f*%s",
        right_ch_name,
        math.max(0,math.min(1,right*2)), right_ch_name)
end
```

**Reply by hooke007** [reply]
精度低了，低步进的值不变化
![Snipaste_2023-08-13_14-05-48](https://github.com/hooke007/MPV_lazy/assets/41094733/39a8160f-6579-470c-a2ca-9ab397ff16ae)

**Reply by YSYHZ** [reply]
![image](https://github.com/hooke007/MPV_lazy/assets/108722788/ddeaf064-83a0-4cac-a4a0-d334b231a71d)
![image](https://github.com/hooke007/MPV_lazy/assets/108722788/c9b54d84-1fdb-4ba7-afdf-d81099ed955a)
我这边是按一下精度还是5%
精度是由下面change_balance控制的，我没改下面的参数
```
mp.add_key_binding("(", 'balance-to-left', function() change_balance(-0.05); end, { repeatable = true })
mp.add_key_binding(")", 'balance-to-right', function() change_balance(0.05); end, { repeatable = true })
```

**Reply by hooke007** [reply]
我知道原始步进精度低，我指出的是平衡的计算精度低

**Reply by YSYHZ** [reply]
懂了，感谢指导
